demo/fiber/Ring_test/src/main/java/org/example/Ring.java

```diff
@@ -0,0 +1,171 @@
+
+/*
+ * Copyright (c) 2021, 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.example;
+
+import org.openjdk.jmh.annotations.*;
+import org.openjdk.jmh.results.format.ResultFormatType;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+import java.util.concurrent.*;
+import java.util.function.Predicate;
+
+
+@Fork(3)
+@Warmup(iterations = 20, time = 1)
+@Measurement(iterations = 20, time = 1)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@BenchmarkMode(Mode.AverageTime)
+@State(Scope.Thread)
+@SuppressWarnings("preview")
+public class Ring {
+
+	public static class Worker<T> implements Runnable {
+		private final Channel<T> source;
+		private final Channel<T> sink;
+		private final Predicate<T> finisher;
+
+		public Worker(Channel<T> source, Channel<T> sink, Predicate<T> finisher) {
+			this.source = source;
+			this.sink = sink;
+			this.finisher = finisher;
+		}
+
+		@Override
+		public void run() {
+			boolean endOfWork = false;
+			do {
+				T msg = source.receive();
+				endOfWork = finisher.test(msg);
+				sink.send(msg);
+			} while (!endOfWork);
+		}
+	}
+
+	@Param({"10"})
+	int THREADS;
+
+	@Param({"sq"})
+	public String queue;
+
+	@Param({"1", "4", "16", "64", "256"})
+	public int stackDepth;
+
+	@Param({"1", "4", "8"})
+	public int stackFrame;
+
+	@Param({"true", "false"})
+	public boolean allocalot;
+
+	@Param({"true", "false"})
+	public boolean singleshot;
+
+	@Setup
+	@SuppressWarnings("unchecked")
```

Comment:
Java Generics is a bit painful.
You cannot create generic arrays.
However, you may try to use `List<Worker<Integer>>` instead of `Worker<Integer>[]`.